### PR TITLE
Refactoring of using ExternalLoader in dictionary DDL.

### PR DESCRIPTION
**Pull request by vitlibar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category:
- Non-significant (changelog entry is not needed)

Instead of using ExternalLoader::reload() now it's used reloadConfig() which reloads only what necessary. Functions attachDictionary() and detachDictionary() are simplified and have lesser number of parameters. Instead of injecting into LoadablesConfigReader's internals for creating dictionary a temp repository is used.

### CODE REVIEW

**Review by alexey-milovidov — APPROVED** [review]

File: dbms/src/Databases/DatabasesCommon.cpp [hunk 1]
```diff
@@ -184,22 +186,19 @@ void DatabaseWithOwnTablesBase::attachTable(const String & table_name, const Sto
...
     String full_name = getDatabaseName() + "." + dictionary_name;
     {
         std::lock_guard lock(mutex);
-        auto status = external_loader.getCurrentStatus(full_name);
-        if (status != ExternalLoader::Status::NOT_EXIST || !dictionaries.emplace(dictionary_name).second)
-            throw Exception(
-                      "Dictionary " + full_name + " already exists.",
-                      ErrorCodes::DICTIONARY_ALREADY_EXISTS);
+        if (!dictionaries.emplace(dictionary_name).second)
+            throw Exception("Dictionary " + full_name + " already exists.", ErrorCodes::DICTIONARY_ALREADY_EXISTS);
```
Comment (alesapin) [inline dbms/src/Databases/DatabasesCommon.cpp]: Why we removed this check?

File: dbms/src/Dictionaries/getDictionaryConfigurationFromAST.cpp [hunk 2]
```diff
@@ -427,7 +427,8 @@ DictionaryConfigurationPtr getDictionaryConfigurationFromAST(const ASTCreateQuer
 
     AutoPtr<Poco::XML::Element> name_element(xml_document->createElement("name"));
     current_dictionary->appendChild(name_element);
-    AutoPtr<Text> name(xml_document->createTextNode(query.database + "." + query.table));
+    String full_name = (!database_name.empty() ? database_name : query.database) + "." + query.table;
```
Comment (alesapin) [inline dbms/src/Dictionaries/getDictionaryConfigurationFromAST.cpp]: Seems like it's already substituted in InterpreterCreateQuery https://github.com/clickhouse/ClickHouse/blob/master/dbms/src/Interpreters/InterpreterCreateQuery.cpp#L714.

File: dbms/src/Interpreters/ExternalLoaderPresetConfigRepository.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+#include <Core/Types.h>
+#include <unordered_map>
+#include <Interpreters/IExternalLoaderConfigRepository.h>
+#include <Poco/Timestamp.h>
+
+
+namespace DB
+{
+/// A config repository filled with preset loadables used by ExternalLoader.
```
Comment (alesapin) [inline dbms/src/Interpreters/ExternalLoaderPresetConfigRepository.h]: This comment doesn't help.

File: dbms/src/Databases/DatabasesCommon.cpp (hunk 1 above)
Comment (vitlibar) [reply dbms/src/Databases/DatabasesCommon.cpp]: `createDictionary()` adds a temp metadata file with a dictionary to `ExternalLoader`, then loads it (if `lazy_load == false`), then calls `attachDictionary()`. At the moment when `attachDictionary()` is called the dictionary could be already loaded. I don't think this check was very useful before anyway.

File: dbms/src/Dictionaries/getDictionaryConfigurationFromAST.cpp (hunk 2 above)
Comment (vitlibar) [reply dbms/src/Dictionaries/getDictionaryConfigurationFromAST.cpp]: It's substituted, but the result of that substitution (i.e. database's name) isn't passed to `createDictionary()`. Seems it's better to set `create.database` [there](https://github.com/clickhouse/ClickHouse/blob/master/dbms/src/Interpreters/InterpreterCreateQuery.cpp#L714) than this solution, I'll simplify the code in a separate PR.

File: dbms/src/Interpreters/ExternalLoaderPresetConfigRepository.h (hunk 3 above)
Comment (vitlibar) [reply dbms/src/Interpreters/ExternalLoaderPresetConfigRepository.h]: I think I'll rename this class to something like `ExternalLoaderTempConfigRepository` and write a decent comment in a separate PR.

**Pull request merged**